torch/testing/_internal/common_optimizers.py

```diff
@@ -0,0 +1,626 @@
+                            )
+                        ]
+                    }
+                ],
+                kwargs={},
+                desc="dense params required in param_groups",
+            ),
+            error_type=ValueError,
+            error_regex="SparseAdam requires dense parameter tensors",
+        ),
+    ]
+
+
+# Database of OptimizerInfo entries in alphabetical order.
+optim_db: List[OptimizerInfo] = [
+    OptimizerInfo(
+        Adadelta,
+        optim_inputs_func=None,
+        optim_error_inputs_func=optim_error_inputs_func_adadelta,
+        supported_impls=("foreach", "differentiable"),
+    ),
+    OptimizerInfo(
+        Adagrad,
+        optim_inputs_func=None,
+        optim_error_inputs_func=optim_error_inputs_func_adagrad,
+        supported_impls=("foreach", "differentiable"),
+        supports_sparse_on=("cpu"),
+    ),
+    OptimizerInfo(
+        Adam,
+        optim_inputs_func=None,
+        optim_error_inputs_func=optim_error_inputs_func_adam,
+        supported_impls=("foreach", "differentiable", "fused"),
+    ),
+    OptimizerInfo(
+        Adamax,
+        optim_inputs_func=None,
+        optim_error_inputs_func=optim_error_inputs_func_adamax,
+        supported_impls=("foreach", "differentiable"),
+    ),
+    OptimizerInfo(
+        AdamW,
+        optim_inputs_func=None,
+        optim_error_inputs_func=optim_error_inputs_func_adamw,
+        supported_impls=("foreach", "differentiable", "fused"),
+    ),
+    OptimizerInfo(
+        ASGD,
+        optim_inputs_func=None,
+        optim_error_inputs_func=optim_error_inputs_func_asgd,
+        supported_impls=("foreach", "differentiable"),
+    ),
+    OptimizerInfo(
+        LBFGS,
+        optim_inputs_func=None,
+        optim_error_inputs_func=optim_error_inputs_func_lbfgs,
+        supported_impls=(),
+        step_requires_closure=True,
+        supports_param_groups=False,
+        supports_multiple_devices=False,
+    ),
+    OptimizerInfo(
+        NAdam,
+        optim_inputs_func=None,
+        optim_error_inputs_func=optim_error_inputs_func_nadam,
+        supported_impls=("foreach", "differentiable"),
+    ),
+    OptimizerInfo(
+        RAdam,
+        optim_inputs_func=None,
+        optim_error_inputs_func=optim_error_inputs_func_radam,
+        supported_impls=("foreach", "differentiable"),
+    ),
+    OptimizerInfo(
+        RMSprop,
+        optim_inputs_func=None,
+        optim_error_inputs_func=optim_error_inputs_func_rmsprop,
+        supported_impls=("foreach", "differentiable"),
+    ),
+    OptimizerInfo(
+        Rprop,
+        optim_inputs_func=None,
+        optim_error_inputs_func=optim_error_inputs_func_rprop,
+        supported_impls=("foreach", "differentiable"),
+    ),
+    OptimizerInfo(
+        SGD,
+        optim_inputs_func=None,
+        optim_error_inputs_func=optim_error_inputs_func_sgd,
+        supported_impls=("foreach", "differentiable"),
+        supports_sparse_on=("cpu", "cuda"),
```

Comment:
Given that the test we have only runs on cpu, the cuda here does nothing in this PR and will be used in a follow above I guess?

Reply:
yes! :D